### PR TITLE
Team 7 animals into main

### LINKED ISSUE #454: Animals-scaling
# **Description**
**Task:** [Animals Scaling]().
This task was to implement animal sprite scaling + density scaling (bosses should not be easily moveable). 
 
# **Milestones**
List of steps that need to be completed for this task.
- [x] Add scaling code into factories and configs (Oct 8)
- [x] Add scaling parameters to configs (Oct 10)
- [x] Add density code to factories (Oct 11)
- [x] Add density parameters to configs (Oct 12)
- [x] Add above to pets (Oct 12)
 
**Completion Deadline:** Oct 12

# **Documentation**
 
# **Member**
- Varun sinha @vsinh14

### PR DESCRIPTION
**Pull request by Aryan-S-Gupta**
# Description

This Pull Request implements several significant enhancements to our game, focusing on improving boss mechanics, refactoring key components, and adding new features. The changes include:

1. Implementation of the "BulletStorm" feature for boss encounters, enhancing gameplay depth and challenge.
2. Introduction of a new final boss, Cthulhu, replacing Birdman (who now replaces Werewolf as the first boss).
3. Refactoring of NPCFactory, NPCConfigs, and PetFactory for improved maintainability.
4. Decoupling of the Effects System and implementation of animated overlays.
5. Refactoring and extension of RangeAttackComponent for the Final Boss.
6. Implementation of animal sprite scaling and density scaling.
7. Addition of new dialogue lines for Boss characters.

These changes significantly enhance player engagement, game depth, and overall code quality.

Closes #454 #430 #404 #427 #439 #440

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# How Has This Been Tested?

The following tests have been performed to verify the changes:

- [x] Basic NPCConfigsTest: Validates the configuration of NPCs.
- [x] WeaponFactoryTest: Fixed and verified functionality.
- [x] NPCDeathHandler Test: Basic test case to verify behavior and correctness.
- [x] Comprehensive testing of BulletStorm mechanics across both Kitsune and Cthulhu bosses.
- [x] Verification of animal sprite scaling and density scaling.
- [x] Testing of new dialogue lines for Boss characters.

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: source/core/src/main/com/csse3200/game/areas/MainRoom.java [hunk 1]
```diff
@@ -7,10 +7,9 @@
 import java.util.List;
 
 public class MainRoom extends EnemyRoom {
-    @Override
-    protected List<List<String>> getAnimalSpecifications() {
+    @Override protected List<List<String>> getAnimalSpecifications() {
         return List.of(
-                List.of("Dog", "Snake"),
+                List.of("Rat", "Bear", "Bat", "Dog", "Snake", "Dino", "Minotaur","Dragon"),
```
Comment (manya-k) [inline source/core/src/main/com/csse3200/game/areas/MainRoom.java]: Reset this list

**Review by manya-k — APPROVED** [review]
code working as expected. the bugs are fixed. 
game balance needs improvement but can be done In another pr

File: source/core/src/main/com/csse3200/game/areas/MainRoom.java (hunk 1 above)
Comment (manya-k) [reply source/core/src/main/com/csse3200/game/areas/MainRoom.java]: done

**Comment by manya-k** [comment]
after loading, the boss is out of the map. pls fix the bug

**Review by dhruvanil55 — COMMENTED** [review]
The pr is buggy, boss spawns outside of the map.

**Comment by manya-k** [comment]
reviewed the changes, working as expected.
ready to merge

**Review by dhruvanil55 — APPROVED** [review]

**Comment by dhruvanil55** [comment]
> The pr is buggy, boss spawns outside of the map.

FIXED TEAM 6 bugs

**Pull request merged**